include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -33,6 +38,88 @@ namespace execution
 inline namespace __dpl
 {
 
+namespace __internal
+{
+
+#if _ONEDPL_PREDEFINED_POLICIES
+struct __global_instance_tag {};
+#endif
+
+class alignas(sycl::queue) __queue_holder
+{
+    static_assert(sizeof(sycl::queue) >= sizeof(void*));
+    static_assert(alignof(sycl::queue) >= alignof(void*));
+
+    std::byte __buf[sizeof(sycl::queue)];
+
+    bool __has_queue() const
+    {
+        bool res = (nullptr != reinterpret_cast<void* const&>(*this));
+        std::atomic_signal_fence(std::memory_order_acq_rel); // mitigate possible reordering due to type punning
+        return res;
+    }
+
+  public:
+    template <typename... _Args>
+    __queue_holder(_Args... __args)
+    {
+        new (this) sycl::queue(__args...);
+    }
+
+#if _ONEDPL_PREDEFINED_POLICIES
+    __queue_holder(__global_instance_tag)
+    {
+        if (!sycl::device::get_devices().empty())
+            new (this) sycl::queue;
+        else
+        {
+            // an "impossible" case of SYCL providing no devices, which we however must handle
+            std::memset(this, 0, sizeof(void*));
+            // Since the queue holder does not have a valid queue in this case,
+            // the predefined device policies are de-facto unusable.
+            // That seems OK though, because there is no device to run SYCL anyway.
+        }
+    }
+#endif
+
+    ~__queue_holder()
+    {
+        if (__has_queue())
+            __queue_ref().~queue();
+    }
+
+    // Copy and move operations have to be explicit
+    __queue_holder(const __queue_holder& __h) { new (this) sycl::queue(__h.__queue_ref()); }
+    __queue_holder(__queue_holder&& __h)      { new (this) sycl::queue(std::move(__h.__queue_ref())); }
+
+    __queue_holder& operator=(const __queue_holder& __h)
+    {
+        if (this != &__h)
+            __queue_ref() = __h.__queue_ref();
+        return *this;
+    }
+
+    __queue_holder& operator=(__queue_holder&& __h)
+    {
+        if (this != &__h)
+            __queue_ref() = std::move(__h.__queue_ref());
+        return *this;
+    }
+
+    const sycl::queue& __queue_ref() const
+    {
+        assert(__has_queue());
+        return reinterpret_cast<const sycl::queue&>(*this);
+    }
```

Comment:
With the assertion in `__queue_ref()`, and the use of `__queue_ref()` within assignment operators, copy and move constructors, we are basically saying if no device is available, you cannot move or copy a `__queue_holder` (with runtime assertions enabled) and therefore the same would go for the defaulted policies themselves.  
In principle, we could handle this case by moving / copying the zeroed out bytes.  

Is the idea that if no device is available, we shouldn't be doing anything the default policy at all? 
I suppose there is a small benefit in avoiding a branch as written by not supporting the "no device" case.

However, a user may believe that as long as they don't do something like call an algorithm with a policy, they are free to copy/move it around even without a device present.

Reply:
In order to copy or move a policy, it would first need to be created. But the creation will fail, except for the predefined policies. Therefore I think caring of copy/move operations for invalid policies is not worth the trouble.

Reply:
So... how about a case like this: 
``` 
#include <oneapi/dpl/execution>
#include <oneapi/dpl/algorithm>
#include <sycl/sycl.hpp>

template <typename Policy>
void
function(Policy policy)
{
	std::vector<int> v{10,0};

	if (!sycl::device::get_devices().empty())
	{
		std::sort(policy, v.begin(), v.end());
	}
	else 
	{
		std::sort(v.begin(), v.end());
	}
}

int main()
{
	function(oneapi::dpl::execution::dpcpp_default);

    return 0;
}
```

Admittedly, I'm not sure this use case has much merit, but this is a case I could conceive of.  

A user has some code which can determine if its reasonable to call a parallel API based on if valid devices exist, but it is buried in their code somewhere.  At runtime, this would result in a valid serial sort being called when no device is available if no assertion was triggered.  With this branch, the assertion is triggered on copy of the policy into the function.  

Of course it could be argued that a user should just use `oneapi::dpl::execution::dpcpp_default` directly rather than passing it around, but I just want to make sure we are fully solving the problem we want to be solving with this fix.

If in these places we check `__has_queue()` first and `memset` to `0` in the `else`, we could run this code without issue.

Reply:
Passing the policy by reference, as in `template <typename Policy> void function(Policy& policy)`, would do the thing equally well, right? And still no other policy besides the default one can be used with `function`.

At some extra cost, we can have a more elaborated and perhaps more robust solution; e.g. the initial version of this patch tried to be somewhat more generic: https://github.com/oneapi-src/oneDPL/compare/main...b83ab2d
But I am not sure it makes much sense to do that way. If we want/need to have a better solution, I think we should bite the bullet and reimplement the class with real lazy-init behavior for all instances, accepting and properly handling backward incompatibility.

Reply:
Yes, passing by reference resolves the issue in the example, and that is the norm for passing around policies.
I'm probably making an issue where one doesn't exist in practice, or is at least easily worked around. 

Once we are ready to merge, it might be good to reach out to the reporter of #1060 to confirm that this branch resolves the issue fully for their case, as it is an active example of someone experiencing this problem.

Reply:
Yes, I agree.  This is clearly outside the realm of the SYCL spec, and we can have a clear solution for users if they still encounter problems:  
If there is no device, you must avoid touching (copying, moving, calling APIs, ...)  `dpcpp_default` or `default_selector_v`.